### PR TITLE
SpectralLibraryGenerationFirstAttempt

### PR DESCRIPTION
**Pull request by Yuling1996**

### CODE REVIEW

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 1]
```diff
@@ -741,7 +741,8 @@ public static PsmData CreateOnePsmDataEntry(string searchType, List<(string file
             float isLoop = 0;
             float isInter = 0;
             float isIntra = 0;
-            float spectralAngle = 0;
+            float spectralAngle = (float)double.NaN;
+            float hasSpectralAngle = 0;
```
[the remark this answers is not in the document]
Comment (Yuling1996) [reply EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs:745]: hasSpectralAngle will be used in IImmutableDictionary<string, int> assumedAttributeDirection, so I think it need to be a number. Did I misunderstand something ?

File: TaskLayer/SearchTask/PostSearchAnalysisTask.cs [hunk 2]
```diff
@@ -580,6 +585,98 @@ private void WritePsmResults()
...
+
+                    // sort psms by MetaMorpheus scores, the first one has the highest  MetaMorpheus score. Take top 5 spectra and compare each one with other 4 spectra and calculate 
+                    // spectral angles. The one which has largest sum of spectral angles will be treated as initial standard spectrum.
+                    List<double> eachSaTotal = new List<double>();
+                    while (b < 5 && b < x.Value.Count)
+                    {
+                        int c = 0;
+                        List<MatchedFragmentIon> possibleStandardPeaks = x.Value[b].MatchedFragmentIons;
+                        List<double> saList = new List<double>();
+                        {
```
[the remark this answers is not in the document]
Comment (Yuling1996) [reply TaskLayer/SearchTask/PostSearchAnalysisTask.cs:636]: Yes,  a line of code was missing here. I will fix it soon

File: EngineLayer/SpectralLibrarySearch/SpectralLibrarySearchFunction.cs [hunk 3]
```diff
@@ -73,27 +80,131 @@ public static double CalculateNormalizedSpectralAngle(List<MatchedFragmentIon> t
...
                 }
             }
 
             double normalizedSpectralAngle = 1 - (2 * Math.Acos(dotProduct) / Math.PI);
 
             return normalizedSpectralAngle;
         }
+
+        // For decoy library spectrum generation, we use the predicted m/z valuse of the decoy sequence and we use the decoy's corresponding target's library spectrum's intensity values as decoy's intensities
+        public static List<MatchedFragmentIon> GetDecoyLibrarySpectrumFromTargetByReverse(LibrarySpectrum targetSpectrum, List<Product> decoyPeptideTheorProducts)
```
[the remark this answers is not in the document]
Comment (Yuling1996) [reply EngineLayer/SpectralLibrarySearch/SpectralLibrarySearchFunction.cs:110]: Does this function make sense: TestDecoyLibrarySpectraGeneration() ?

**Review by kyp4 — CHANGES_REQUESTED** [review]
please make it so that you cannot run MM with only a library when it requires a library and a fasta to properly run a search

**Comment by Yuling1996** [comment]
> please make it so that you cannot run MM with only a library when it requires a library and a fasta to properly run

**Comment by Yuling1996** [comment]
> > please make it so that you cannot run MM with only a library when it requires a library and a fasta to properly run

**Comment by Yuling1996** [comment]
> > > please make it so that you cannot run MM with only a library when it requires a library and a fasta to properly run

I will make changes to this. Thanks

File: EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs [hunk 4]
```diff
@@ -900,6 +907,16 @@ private static bool PeptideIsVariant(PeptideWithSetModifications pwsm)
             return identifiedVariant;
         }
 
+        private static bool PsmHasSpectralAngle(PeptideSpectralMatch psm)
+        {
```
[the remark this answers is not in the document]
Comment (Yuling1996) [reply EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs:912]: solved

File: EngineLayer/SpectralLibrarySearch/SpectralLibrarySearchFunction.cs [hunk 5]
```diff
@@ -22,22 +22,29 @@ public static class SpectralLibrarySearchFunction
                 //TODO: spectral angle could be used to disambiguate PSMs. right now for ambiguous PSMs, the spectral angle for only one peptide option is saved
                 foreach (var peptide in psm.PeptidesToMatchingFragments)
                 {
-                    if (spectralLibrary == null || !spectralLibrary.TryGetSpectrum(peptide.Key.FullSequence, scan.PrecursorCharge, out var librarySpectrum))
+                    //if peptide is target, directly look for the target's spectrum in the spectral library
+                    if (spectralLibrary != null && !peptide.Key.Protein.IsDecoy && spectralLibrary.TryGetSpectrum(peptide.Key.FullSequence, scan.PrecursorCharge, out var librarySpectrum))
```
[the remark this answers is not in the document]
Comment (Yuling1996) [reply EngineLayer/SpectralLibrarySearch/SpectralLibrarySearchFunction.cs]: solved

**Comment by rmmiller22** [comment]
New PR was generated for spectral library content making this PR out of date

**Pull request closed without merging**